presto-main/src/main/java/com/facebook/presto/cost/HistoryBasedPlanStatisticsTracker.java

```diff
@@ -133,16 +136,23 @@ public Map<PlanNodeWithHash, PlanStatisticsWithSourceInfo> getQueryStats(QueryIn
                 if (planNodeStats == null) {
                     continue;
                 }
+                double outputPositions = planNodeStats.getPlanNodeOutputPositions();
+                double outputBytes = planNodeStats.getPlanNodeOutputDataSize().toBytes();
+                // HashGenerationOptimizer is only inserted at end, but earlier parts of optimizer use these stats,
+                // and are not aware of these variables. So we remove them from stats calculation.
+                outputBytes -= planNode.getOutputVariables().stream()
+                        .mapToDouble(variable -> variable.getName().startsWith(HASH_VALUE_PREFIX) && variable.getType() instanceof BigintType ? outputPositions * BigintType.BIGINT.getFixedSize() : 0)
+                        .sum();
```

Comment:
This also looks very specialized; would love to get @rschlussel 's feedback.

Reply:
I'm not sure what to think about this, but i lean towards not special casing here
It is true that we do know about this extra field that we would be including even though it hasn't been added yet.  However,
1.  the presence of the hash variable doesn't seem likely to make a big difference in our decisions, so not sure it's worth excluding it
2. even though it hasn't been added yet, it actually will be added later (unless someone turns off the hash generation optimizer), so maybe the optimizers would actually be better off if this info were included.  
3. there can be other e.g. adding/removing projections, etc. that happen after stats equivalent plan node, don't want to have special cases for everything.

Reply:
Hmm, based on my tests, this adjustment is needed for some pipelines. If we are just selecting an id, having another variable of same size can lead to 2x differences in stats. 

I understand that of course some things will be excluded anyway, but this heuristic is sufficient for most of the queries to give precise statistics.

However, I changed the code to do some generic adjustments by comparing the variables between plan node and stats equivalent plan node, and looking for differences.